### ISSUE TITLE
HTTPErrors when making Bugzilla API requests result in a ValueError

### ISSUE THREAD

**Issue by edmorley** [AutomatedTester/Bugsy#27]
There are three main buckets of errors that Bugsy could experience when interacting with Bugzilla's REST API:
1) Network problem (e.g. DNS failure, refused connection, timeout etc)
2) A non-200 HTTP response that is valid JSON and has "error" and "message" keys (eg Bugzilla responding saying a required field was omitted when creating a bug).
3) A non-200 HTTP response that is not valid JSON (eg server HTTP 500, 404 page, maintenance page, load balancer error page etc)

Currently (1) is handled fine, since requests raises the correct exception itself - eg:
```python
bz = bugsy.Bugsy(bugzilla_url="http://invalid-url")
bz.get(123456)
Traceback (most recent call last):
    ...
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', gaierror(8, 'Name or service not known'))
```
And as of #24 and #25, Bugsy also handles (2) fine, since it now returns a BugsyException with relevant message from the JSON error response "message" key.

However (3) still results in unexpected behaviour - ie for all of the following:
* Requests to verify login
* Creating/updating a bug
* Fetching a bug
* Adding a comment to a bug

...if an HTTP 500 is experienced for the request, then Bugsy currently returns `ValueError: No JSON object could be decoded`, since it doesn't use `.json()` within a try-except. This means the user doesn't see the original cause, which was an HTTPError.

Also for the `add_comment()` case, if the HTTP response happened to be something that the JSON parser would accept (eg ``body='"Foo"'``), then the request fails silently even though it wasn't an HTTP 200 response.

The best fix for this is likely to use `raise_for_status()` (as mentioned in #25) as a fallback after checking if an `error`/`message` key exists (ie to make sure we don't regress the (2) case above).

I have some tests that currently fail that demonstrate the issue - I'll open a PR to add them as expected fail, which will hopefully make this clearer :-) If I have time I'll also then open a PR to fix them (and remove the xfail annotation).

**Comment by AutomatedTester** [comment]
I am sorry, I didnt notice you referenced a PR in this bug but have fixed this now in a4c1105f3d16bc6279365774fda127136c2e83bf

**Comment by edmorley** [comment]
Thank for landing some fixes - however unfortunately:
* the `add_comment()` case in the previous PR (#28) still fails
* HTTP 500s during adding or removing tags are now silently dropped
* other HTTP errors (eg 502, 503) aren't handled (HTTP 500 was just an example, I didn't mean for it to be special-cased specifically)

I've opened another PR containing currently-failing testcases since it's probably the quickest way to explain the above - #30 - the travis log for it is here:
https://travis-ci.org/AutomatedTester/Bugsy/builds/83117760

That said, it's up to you whether you fix, I've run out of time to contribute to Bugsy & the OrangeFactor script is using the "add comment" API directly so I can handle these cases & also support using requests retries feature (via a custom HTTP adapter) etc. (Many of the advantages Bugsy gives aren't needed for the script since all it does is add comments to existing bugs.)
